addons/tls13-middlebox/bootstrap.js

```diff
@@ -0,0 +1,241 @@
+const FALLBACK_LIMIT_PREF = "security.tls.version.fallback-limit";
+
+let readwrite_prefs = new Preferences({defaultBranch: true});
+
+// all combination of configurations we care about.
+let configurations = [
+  {maxVersion: 4, fallbackLimit: 4, website: "enabled.tls13.com"},
+  {maxVersion: 4, fallbackLimit: 4, website: "disabled.tls13.com"},
+  {maxVersion: 3, fallbackLimit: 3, website: "control.tls12.com"}
+];
+
+// some fields are not available sometimes, so we have to catch the errors and return undefined.
+function getFieldValue(obj, name) {
+  try {
+    return obj[name];
+  } catch (ex) {
+    return undefined;
+  }
+}
+
+// extracting the root certificate from the chain
+function extractRootCert(cert) {
+  let root_cert = cert;
+
+  while (getFieldValue(root_cert, "issuer")) {
+    root_cert = getFieldValue(root_cert, "issuer");
+  }
+
+  return root_cert;
+}
+
+// returns true if there is at least one non-builtin root certificate is installed
+function isNonBuiltInRootCertInstalled() {
+  let certDB = Cc["@mozilla.org/security/x509certdb;1"].getService(Ci.nsIX509CertDB);
+
+  let iter = certDB.getCerts().getEnumerator();
+
+  while (iter.hasMoreElements()) {
+    let cert = iter.getNext().QueryInterface(Ci.nsIX509Cert);
+
+    if (getFieldValue(cert, "issuer") === null &&
+      getFieldValue(cert, "isBuiltInRoot") === false &&
+      // There are some root certificates that are built-in but their isBuiltInRoot is false.
+      // That is why we check their tokenName as well
+      getFieldValue(cert, "tokenName").toLowerCase() !== "Builtin Object Token".toLowerCase()) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+function getInfo(xhr) {
+  let result = {};
+
+  try {
+    let channel = xhr.channel;
+
+    // this is the most important value based on which we can find out the problem
+    channel.QueryInterface(Ci.nsIRequest);
+    result.status = getFieldValue(channel, "status");
+
+    let securityInfo = getFieldValue(channel, "securityInfo");
+
+    if (securityInfo instanceof Ci.nsITransportSecurityInfo) {
+      securityInfo.QueryInterface(Ci.nsITransportSecurityInfo);
+
+      result.securityState = getFieldValue(securityInfo, "securityState");
+
+      // Error message on connection failure. I am not sure if we can get this error message using status code.
+      // It is safer to collect this information as well.
+      result.errorMessage = getFieldValue(securityInfo, "errorMessage");
+    }
+
+    if (securityInfo instanceof Ci.nsISSLStatusProvider) {
+      securityInfo.QueryInterface(Ci.nsISSLStatusProvider);
+      let sslStatus = securityInfo.SSLStatus;
+
+      if (sslStatus) {
+        sslStatus.QueryInterface(Ci.nsISSLStatus);
+
+        // extracting sha256 fingerprint for the leaf cert
+        result.serverCertSha256Fingerprint = getFieldValue(sslStatus.serverCert, "sha256Fingerprint");
+
+        let root_cert = extractRootCert(sslStatus.serverCert);
```

Comment:
I believe if certificate verification failed, sslStatus.serverCert won't be set. In that case, you should use securityInfo.failedCertChain.